File: backend/users/serializers/subscriptions.py
"""
Serializers for subscription management.
"""
from rest_framework import serializers
from users.models.subscriptions import (
    ClientSubscription,
    SubscriptionPreference,
    SubscriptionType,
    DeliveryChannel
)


class ClientSubscriptionSerializer(serializers.ModelSerializer):
    """Serializer for client subscriptions."""
    subscription_type_display = serializers.CharField(
        source='get_subscription_type_display',
        read_only=True
    )
    
    class Meta:
        model = ClientSubscription
        fields = [
            'id',
            'subscription_type',
            'subscription_type_display',
            'is_subscribed',
            'subscribed_at',
            'unsubscribed_at',
            'frequency',
            'preferred_channels',
            'last_sent_at',
            'created_at',
            'updated_at',
        ]
        read_only_fields = [
            'id', 'subscribed_at', 'unsubscribed_at', 
            'last_sent_at', 'created_at', 'updated_at'
        ]


class SubscriptionPreferenceSerializer(serializers.ModelSerializer):
    """Serializer for subscription preferences."""
    
    class Meta:
        model = SubscriptionPreference
        fields = [
            'id',
            'all_subscriptions_enabled',
            'marketing_consent',
            'marketing_consent_date',
            'email_enabled',
            'sms_enabled',
            'push_enabled',
            'in_app_enabled',
            'dnd_enabled',
            'dnd_start_hour',
            'dnd_end_hour',
            'transactional_enabled',
            'created_at',
            'updated_at',
        ]
        read_only_fields = ['id', 'marketing_consent_date', 'created_at', 'updated_at']


class SubscriptionListSerializer(serializers.Serializer):
    """Serializer for listing all available subscriptions."""
    subscription_type = serializers.ChoiceField(choices=SubscriptionType.choices)
    is_subscribed = serializers.BooleanField()
    subscribed_at = serializers.DateTimeField(required=False, allow_null=True)
    unsubscribed_at = serializers.DateTimeField(required=False, allow_null=True)
    frequency = serializers.CharField()
    preferred_channels = serializers.ListField(
        child=serializers.ChoiceField(choices=DeliveryChannel.choices),
        required=False
    )


class SubscribeRequestSerializer(serializers.Serializer):
    """Serializer for subscription requests."""
    subscription_type = serializers.ChoiceField(choices=SubscriptionType.choices)
    frequency = serializers.ChoiceField(
        choices=[
            ('immediate', 'Immediate'),
            ('daily', 'Daily Digest'),
            ('weekly', 'Weekly Digest'),
            ('monthly', 'Monthly Digest'),
        ],
        default='immediate',
        required=False
    )
    preferred_channels = serializers.ListField(
        child=serializers.ChoiceField(choices=DeliveryChannel.choices),
        required=False,
        allow_empty=True
    )


class UnsubscribeRequestSerializer(serializers.Serializer):
    """Serializer for unsubscription requests."""
    subscription_type = serializers.ChoiceField(choices=SubscriptionType.choices)


class UpdateFrequencySerializer(serializers.Serializer):
    """Serializer for updating subscription frequency."""
    subscription_type = serializers.ChoiceField(choices=SubscriptionType.choices)
    frequency = serializers.ChoiceField(
        choices=[
            ('immediate', 'Immediate'),
            ('daily', 'Daily Digest'),
            ('weekly', 'Weekly Digest'),
            ('monthly', 'Monthly Digest'),
        ]
    )


class UpdateChannelsSerializer(serializers.Serializer):
    """Serializer for updating preferred channels."""
    subscription_type = serializers.ChoiceField(choices=SubscriptionType.choices)
    preferred_channels = serializers.ListField(
        child=serializers.ChoiceField(choices=DeliveryChannel.choices),
        allow_empty=True
    )

File: frontend/src/api/subscriptions.ts
import { api, apiPath } from "./client";

const base = (path = "") => apiPath(`/users/subscriptions${path}`);

export type SubscriptionType =
  | "newsletter"
  | "blog_posts"
  | "coupon_updates"
  | "marketing_messages"
  | "unread_messages"
  | "transactional_messages"
  | "notifications"
  | "order_updates"
  | "promotional_offers"
  | "product_updates"
  | "security_alerts"
  | "account_updates";

export type DeliveryChannel = "email" | "sms" | "push" | "in_app";
export type Frequency = "immediate" | "daily" | "weekly" | "monthly";

export interface SubscriptionItem {
  subscription_type: SubscriptionType;
  is_subscribed: boolean;
  subscribed_at: string | null;
  unsubscribed_at: string | null;
  frequency: Frequency;
  preferred_channels: DeliveryChannel[];
}

export interface SubscriptionPreferences {
  id?: number;
  all_subscriptions_enabled: boolean;
  marketing_consent: boolean;
  marketing_consent_date: string | null;
  email_enabled: boolean;
  sms_enabled: boolean;
  push_enabled: boolean;
  in_app_enabled: boolean;
  dnd_enabled: boolean;
  dnd_start_hour: number;
  dnd_end_hour: number;
  transactional_enabled: boolean;
  created_at?: string;
  updated_at?: string;
}

export const subscriptionsApi = {
  listAll: () =>
    api.get<SubscriptionItem[]>(base("/list-all/")),

  preferences: () =>
    api.get<SubscriptionPreferences>(base("/preferences/")),

  subscribe: (payload: {
    subscription_type: SubscriptionType;
    frequency?: Frequency;
    preferred_channels?: DeliveryChannel[];
  }) => api.post(base("/subscribe/"), payload),

  unsubscribe: (subscription_type: SubscriptionType) =>
    api.post(base("/unsubscribe/"), { subscription_type }),

  updateFrequency: (subscription_type: SubscriptionType, frequency: Frequency) =>
    api.put(base("/update-frequency/"), { subscription_type, frequency }),

  updateChannels: (subscription_type: SubscriptionType, preferred_channels: DeliveryChannel[]) =>
    api.put(base("/update-channels/"), { subscription_type, preferred_channels }),

  updatePreferences: (data: Partial<SubscriptionPreferences>) =>
    api.put(base("/update-preferences/"), data),
};
